File: js/leetcode/A297_Serialize_and_Deserialize_Binary_Tree.js
const TreeNode = require('./datastructure').TreeNode;

const serialize1 = (root) => {
    let res = '';
    const buildString = (node) => {
        if (!node) {
            res += 'N,';
        } else {
            res += `${node.val},`
            buildString(node.left);
            buildString(node.right);
        }
    }
    buildString(root);
    return res;
}

const deserialize1 = (data) => {
    let queue = data.split(',');
    const buildTree = (strs) => {
        let str = strs.shift();
        if (str === 'N') return null;
        let tn = new TreeNode(~~str);
        tn.left = buildTree(strs);
        tn.right = buildTree(strs);
        return tn;
    }
    return buildTree(queue);
}

File: java/src/leetcode/problems/A297_Serialize_and_Deserialize_Binary_Tree.java
package leetcode.problems;

import leetcode.datastructures.TreeNode;
import leetcode.utils.Serializer;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

public class A297_Serialize_and_Deserialize_Binary_Tree {

    /**
     * Use preorder to serialize and deserialize
     * @param root
     * @return
     */
    public String serialize(TreeNode root) {
        if (root == null) {
            return "";
        }
        List<String> nodesValues = new ArrayList<>();
        serializeHelper(root, nodesValues);
        return String.join(",", nodesValues);
    }

    private void serializeHelper(TreeNode node, List<String> nodesValues) {
        if (node == null) {
            nodesValues.add("#");
            return;
        }
        nodesValues.add(String.valueOf(node.val));
        serializeHelper(node.left, nodesValues);
        serializeHelper(node.right, nodesValues);
    }

    public TreeNode deserialize(String data) {
        if (data.isEmpty()) {
            return null;
        }
        String[] nodesValues = data.split(",");
        int[] index = new int[]{-1};
        return createTreeNode(index, nodesValues);
    }

    private TreeNode createTreeNode(int[] index, String[] nodesValues) {
        index[0]++;
        if (nodesValues[index[0]].equals("#")) {
            return null;
        }
        TreeNode node = new TreeNode(Integer.parseInt(nodesValues[index[0]]));
        node.left = createTreeNode(index, nodesValues);
        node.right = createTreeNode(index, nodesValues);
        return node;
    }

    public static void main(String[] args) {
        Serializer s = new Serializer();
        TreeNode rt = s.deserialize("[1, 2, 3, null, null, 4, 5]");
        A297_Serialize_and_Deserialize_Binary_Tree solution = new A297_Serialize_and_Deserialize_Binary_Tree();

        System.out.println(solution.serialize(solution.deserialize("1,2,#,#,3,4,#,#,5,#,#")));
    }
}
